azext_edge/edge/providers/orchestration/upgrade2.py

```diff
@@ -0,0 +1,309 @@
+            for ext in upgradeable_extensions:
+                updated = self.resource_map.connected_cluster.clusters.extensions.update_cluster_extension(
+                    resource_group_name=self.resource_group_name,
+                    cluster_name=self.resource_map.connected_cluster.cluster_name,
+                    extension_name=ext.extension["name"],
+                    update_payload=ext.get_patch(),
+                    retry_total=0,
+                )
+                return_payload.append(updated)
+                progress.advance(upgrade_task)
+
+            return return_payload
+
+
+def render_upgrade_table(upgrade_state: "ClusterUpgradeState"):
+    table = get_default_table()
+    for ext in upgrade_state.extension_upgrades:
+        patch_payload = ext.get_patch()
+        if not patch_payload:
+            patch_payload = "N/A"
+        else:
+            patch_payload = JSON(dumps(patch_payload))
+
+        table.add_row(
+            f"{ext.moniker}",
+            f"{ext.current_version[0]} \\[{ext.current_version[1]}]",
+            f"{ext.desired_version[0]} \\[{ext.desired_version[1]}]",
+            patch_payload,
+        )
+        table.add_section()
+
+    DEFAULT_CONSOLE.print(table)
+
+
+def build_override_map(**override_kwargs) -> Dict[str, "ConfigOverride"]:
+    result_map = {}
+    for moniker in EXTENSION_MONIKER_TO_ALIAS_MAP:
+        alias = EXTENSION_MONIKER_TO_ALIAS_MAP[moniker]
+        config_override = ConfigOverride(
+            config=override_kwargs.get(f"{alias}_config"),
+            version=override_kwargs.get(f"{alias}_version"),
+            train=override_kwargs.get(f"{alias}_train"),
+        )
+        if not config_override.is_empty:
+            result_map[moniker] = config_override
+
+    return result_map
+
+
+class ConfigOverride:
+    def __init__(
+        self,
+        config: Optional[dict] = None,
+        version: Optional[str] = None,
+        train: Optional[str] = None,
+    ):
+        self.config = parse_kvp_nargs(config)
+        self.version = version
+        self.train = train
+
+    @property
+    def is_empty(self):
+        return not any([self.config, self.version, self.train])
+
+
+class ClusterUpgradeState:
+    def __init__(
+        self,
+        extensions_map: Dict[str, dict],
+        init_version_map: Dict[str, dict],
+        override_map: Dict[str, "ConfigOverride"],
+    ):
+        self.extensions_map = extensions_map
+        self.init_version_map = init_version_map
+        self.override_map = override_map
+        self.extension_upgrades = self.refresh_upgrade_state()
+
+    def has_upgrades(self) -> bool:
+        return any(ext_state.can_upgrade() for ext_state in self.extension_upgrades)
+
+    def refresh_upgrade_state(self) -> List["ExtensionUpgradeState"]:
+        ext_queue: List["ExtensionUpgradeState"] = []
+
+        # TODO @digimaun - deterine further pre-checks.
```

Comment:
are we handling system extensions or not yet (TODO then?)

Reply:
Yea, need input for those decisions. Will include in description.